hipe4ml/TrainingAndTesting/generalized_analysis.py

```diff
@@ -0,0 +1,131 @@
+'''Module containing the class to perform the training and the testing of the models '''
+
+
+import pandas as pd
+import uproot
+from sklearn.model_selection import train_test_split
+
+
+class GeneralizedAnalysis:
+    ''' Class for training and testing machine learning models starting from ROOT TTrees filled with the candidates
+    '''
+
+    def __init__(
+            self, mc_file_name, data_file_name, sig_selection=0, bkg_selection=0,
+            cent_class=((0, 10),
+                        (10, 30),
+                        (30, 50),
+                        (50, 90))):
+
+        self.cent_class = cent_class.copy()
+        self.n_events = [0, 0, 0, 0]
+
+        self.df_signal = uproot.open(mc_file_name)['SignalTable'].pandas.df()
+        self.df_generated = uproot.open(mc_file_name)['GenTable'].pandas.df()
+        self.df_data = uproot.open(data_file_name)['DataTable'].pandas.df()
+
+        self.df_signal['y'] = 1
+        # backup of the data without any selections for the significance
+        self.df_data_bkg = self.df_data.copy()
+        # dataframe for signal and background with preselection
+        if isinstance(sig_selection, str):
+            self.df_signal = self.df_signal.query(sig_selection)
+        if isinstance(bkg_selection, str):
+            self.df_data_bkg = self.df_data_bkg.query(bkg_selection)
+
+        self.hist_centrality = uproot.open(data_file_name)['EventCounter']
+        self.n_events = []
+        for cent in self.cent_class:
+            self.n_events.append(sum(self.hist_centrality[cent[0]+1:cent[1]]))
+
+    def prepare_dataframe(
+            self, training_columns, cent_class=(0, 90), pt_range=(0, 10),
+            ct_range=(0, 100),
+            test=False):
```

Comment:
Also this function is analysis dependent. It should be reworked to be configurable, as an example the preparation could be done taking a string for the query, user defined, to be more flexible.